### PR TITLE
search: fix use of cps_path

### PR DESCRIPTION
**Pull request by dcbaker**
This must be `@prefix@` prefaced, and is used to describe how the prefix is calculated when the cps file is passed by path, not to calculate a prefix which is different than where the cps file is located.

### CODE REVIEW

File: src/cps/search.cpp [hunk 1]
```diff
@@ -309,13 +309,30 @@ namespace cps::search {
...
-                split.pop_back();
+            if (path) {
+                fs::path p{path.value()};
+                if (p.stem() == "") {
+                    p = p.parent_path();
+                }
+                auto f = fs::path{filename};
+                if (f.stem() == "") {
+                    f = f.parent_path();
+                }
```
Comment (mwoehlke) [inline src/cps/search.cpp:323]: Possible off-by-one error? The `cps_path` is the *directory* in which the `.cps` file resides (both per the spec, and in your example), but I think `filename` here is the location of the `.cps` file? I think you need to explicitly always do one iteration of `f = f.parent_path();` before you start comparing.

In fact, if it's the file name, it should never have an empty `.stem()`, correct? In which case, just removing the `if` seems like TRTTD.

**Review by mwoehlke — COMMENTED** [review]
Modulo the one comment (which TBH might be wrong), this looks like an improvement.

However... I would encourage checking [the specification](https://cps-org.github.io/cps/searching.html#prefix-determination) again, because if I'm reading this correctly, it still doesn't line up.

Specifically:
- If `cps_path` is given, it should be honored, period. If you can't resolve the prefix from the given `cps_path`, that `.cps` file should be treated as not-usable.
- The correct order should be:
  1. Honor `cps_path`
  2. Use search prefix
  3. Guess

  Right now, it appears the first two are switched.
- Per your TODO, just looking for "lib" is inadequate and will actually fail on most systems, as most modern systems use either "lib/<arch>" or "lib64". It looks like you also don't correctly implement macOS guessing, though again, I recognize that correct guessing is low priority.

**Comment by dcbaker** [comment]
> Per your TODO, just looking for "lib" is inadequate and will actually fail on most systems, as most modern systems use either "lib/" or "lib64". It looks like you also don't correctly implement macOS guessing, though again, I recognize that correct guessing is low priority.

Right, at the moment whatever happens to work on Windows and MacOS is is either happenstance, or someone else has worked on it. So as long as the Apple CI stays green, for the moment, I think that's fine.

**Comment by dcbaker** [comment]
> Right now, it appears the first two are switched.

The `prefix_path` there is coming from the command line, when a user wants to override the prefix, equivalent to `pkg-config --prefix=...`, otherwise we do prefer the `cps_path`. The override behavior might be something we need to have a wider CPS discussion about.

**Comment by mwoehlke** [comment]
> at the moment whatever happens to work...

Sure. Please don't take my comment as "fix this now!", but rather acknowledging that a) the current behavior needs improvement, and (importantly) b) you're already aware that it does. :slightly_smiling_face:

(Incidentally, I originally wrote `lib/<arch>`, and Markdown thought it was inline HTML; thanks, Markdown. :unamused:)

> The `prefix_path` there is coming from the command line, when a user wants to override the prefix

Ah, okay. That's my ignorance of the code showing, then.

That said... whether it's necessary, or advisable, to allow overriding the `@prefix@` does seem debatable. (Especially as I don't see that `pkg-config` *has* a `--prefix` option? There is `--define-prefix`, but that doesn't seem to let the user say what the prefix *is*.) *If* we allow overrides, however (and they would need to be able to be specified per package!), the override obviously needs to have priority.

Anyway, I *think* it looks right, now, given I've neither tried to test it myself nor looked at the code as a whole. In any case, as I said before, it's an improvement, so IMHO it's fine to land even if more could be done.

**Comment by dcbaker** [comment]
Yeah, we can discuss whether these kind of overrides are useful in general or whether they're an implementation detail of cps-config to be useful as a drop-in for pkg-config. My suspicion is the latter, since the main usecase is for cross compiling where you have a cross environment mounted under a mount path like `/mnt/image/`, so as you build the image piece-by-piece you need the host compatible pkg-config to query `/mnt/image`, but you need to ensure that root compatible flags (like setting rpaths) end up treating `/mnt/image/` as `/`. I'm not sure what we have at the moment is really comparable to what pkgconf does, but I haven't looked to closely at it for a while.

That might be a later discussion for the working group to figure out how useful that is in general.

**Pull request merged**